pkg/proxy/winkernel/proxier.go

```diff
@@ -0,0 +1,1105 @@
+			}
+			for i := range ss.Addresses {
+				addr := &ss.Addresses[i]
+				if addr.IP == "" {
+					glog.Warningf("ignoring invalid endpoint port %s with empty host", port.Name)
+					continue
+				}
+				isLocal := addr.NodeName != nil && *addr.NodeName == hostname
+				epInfo := newEndpointInfo(addr.IP, uint16(port.Port), isLocal)
+				endpointsMap[svcPortName] = append(endpointsMap[svcPortName], epInfo)
+			}
+			if glog.V(3) {
+				newEPList := []*endpointsInfo{}
+				for _, ep := range endpointsMap[svcPortName] {
+					newEPList = append(newEPList, ep)
+				}
+				glog.Infof("Setting endpoints for %q to %+v", svcPortName, newEPList)
+			}
+		}
+	}
+	return endpointsMap
+}
+
+// Translates single Service object to proxyServiceMap.
+//
+// NOTE: service object should NOT be modified.
+func serviceToServiceMap(service *api.Service) proxyServiceMap {
+	if service == nil {
+		return nil
+	}
+	svcName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+	if shouldSkipService(svcName, service) {
+		return nil
+	}
+
+	serviceMap := make(proxyServiceMap)
+	for i := range service.Spec.Ports {
+		servicePort := &service.Spec.Ports[i]
+		svcPortName := proxy.ServicePortName{NamespacedName: svcName, Port: servicePort.Name}
+		serviceMap[svcPortName] = newServiceInfo(svcPortName, servicePort, service)
+	}
+	return serviceMap
+}
+
+// This is where all of the hns -save/restore calls happen.
+// The only other hns rules are those that are setup in iptablesInit()
+// assumes proxier.mu is held
+func (proxier *Proxier) syncProxyRules() {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	start := time.Now()
+	defer func() {
+		SyncProxyRulesLatency.Observe(sinceInMicroseconds(start))
+		glog.V(4).Infof("syncProxyRules took %v", time.Since(start))
+	}()
+	// don't sync rules till we've received services and endpoints
+	if !proxier.endpointsSynced || !proxier.servicesSynced {
+		glog.V(2).Info("Not syncing hns until Services and Endpoints have been received from master")
+		return
+	}
+
+	// We assume that if this was called, we really want to sync them,
+	// even if nothing changed in the meantime. In other words, callers are
+	// responsible for detecting no-op changes and not calling this function.
+	serviceUpdateResult := proxier.updateServiceMap()
+	endpointUpdateResult := proxier.updateEndpointsMap()
+
+	staleServices := serviceUpdateResult.staleServices
+	// merge stale services gathered from updateEndpointsMap
+	for svcPortName := range endpointUpdateResult.staleServiceNames {
+		if svcInfo, ok := proxier.serviceMap[svcPortName]; ok && svcInfo != nil && svcInfo.protocol == api.ProtocolUDP {
+			glog.V(2).Infof("Stale udp service %v -> %s", svcPortName, svcInfo.clusterIP.String())
+			staleServices.Insert(svcInfo.clusterIP.String())
+		}
+	}
+
+	glog.V(3).Infof("Syncing Policies")
+
+	// Program HNS by adding corresponding policies for each service.
+	for svcName, svcInfo := range proxier.serviceMap {
+		if svcInfo.policyApplied {
```

Comment:
This looks kinda odd.  In what cases would we have effectively the same svcInfo object under two different svcName keys in the serviceMap?

Reply:
I'll look into this.

Reply:
@dcbw Where do you see that assumption?